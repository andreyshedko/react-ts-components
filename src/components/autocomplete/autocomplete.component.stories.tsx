import React from 'react';
import { Meta, Story } from '@storybook/react/types-6-0';

import { AutocompleteComponent, AutocompleteComponentProps } from './autocomplete.component';

export default {
    component: AutocompleteComponent,
    title: 'Autocomplete Component',
    parameters: { actions: { argTypesRegex: '^on.*' } }
} as Meta;

const Template: Story<AutocompleteComponentProps> = (args) => <AutocompleteComponent {...args} />;

export const Autocomplete_Component = Template.bind({});
Autocomplete_Component.args = { suggestions: ['Kiev', 'Kirov', 'Moscow', 'Minsk', 'Prague']}