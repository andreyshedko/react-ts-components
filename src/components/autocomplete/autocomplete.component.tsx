import * as React from "react";
import { useReducer } from "react";
import { ErrorBoundary } from "../error.boundary/error.boundary.component";
import { InputComponent } from "../input/input.component";
import { FilteredSuggestionItem, FilteredSuggestions, Main } from "./autocomplete.styled.component";

export interface AutocompleteComponentProps {
  suggestions: string[];
  placeholder?: string;
  onValueSelected: (suggestion: string) => string;
  onChange?: (event: React.ChangeEvent) => React.ChangeEvent;
}

export interface IFilterState {
  filterValue: string;
  filteredValues: string[];
  selectedValue: string;
}

export interface IFilterAction {
  type: string;
  payload?: string | string[];
}

export enum actions {
  setFilter = 'setFilter',
  setFilteredValue = 'setFilteredValue',
  setSelectedValue = 'SetSelectedValue',
  reset = 'reset'
}

export const reducer = (state: IFilterState, action: IFilterAction): IFilterState => {
  switch (action.type) {
    case actions.setFilter:
      return { ...state, filterValue: action.payload as string };
    case actions.setFilteredValue:
      const filteredValues = (action.payload as string[])
        .filter(suggestion =>
          suggestion.toLowerCase().includes(state.filterValue.toLowerCase())
        )
      return { ...state, filteredValues: filteredValues };
    case actions.setSelectedValue:
      return { ...state, selectedValue: action.payload as string }
    case actions.reset:
      return { ...state, filterValue: '', filteredValues: [], selectedValue: '' }
    default:
      return state;
  }
}

export const AutocompleteComponent: React.FC<AutocompleteComponentProps> = (props: AutocompleteComponentProps) => {

  const initialState: IFilterState = { filterValue: '', filteredValues: [], selectedValue: '' };

  const [state, dispatch] = useReducer(reducer, initialState as IFilterState);

  return (
    <ErrorBoundary>
      <Main>
        <InputComponent
          type={"text"}
          name={"autocomplete"}
          required={true}
          placeholder={props.placeholder || ''}
          value={state.filterValue}
          showClearIcon={true}
          onChange={(value) => {
            if (value) {
              dispatch({ type: actions.setFilter, payload: value });
            }
            if (value && value.length > 2) {
              dispatch({ type: actions.setFilteredValue, payload: props.suggestions });
            }
          }}
          onKeyDown={(event) => {
            switch (event.key) {
              case "Backspace": {
                dispatch({ type: actions.reset });
                const currentValue = (event.currentTarget as HTMLInputElement).value;
                dispatch({ type: actions.setFilter, payload: currentValue });
                if (currentValue && currentValue.length > 2) {
                  dispatch({ type: actions.setFilteredValue, payload: props.suggestions });
                } else {
                  dispatch({ type: actions.reset })
                }
                break;
              }
              default: {
                break;
              }
            }
          }}
          onClick={(event) => {
            const currentValue = (event.currentTarget as HTMLInputElement).value;
            if (currentValue && currentValue.length > 2) {
              dispatch({ type: actions.reset });
              dispatch({ type: actions.setFilter, payload: currentValue });
            }
          }}
          onClear={() => {
            dispatch({ type: actions.reset });
          }}
        />
        {state.filteredValues
          && state.filteredValues.length > 0
          && !(state.selectedValue.length > 0) &&
          <FilteredSuggestions>
            {state.filteredValues.map((suggestion, index) =>
              <FilteredSuggestionItem
                tabIndex={0}
                key={index}
                onKeyDown={(event) => {
                  switch (event.key) {
                    case "Enter": {
                      dispatch({ type: actions.reset });
                      dispatch({ type: actions.setSelectedValue, payload: props.suggestions });
                      dispatch({ type: actions.setFilter, payload: suggestion });
                      props.onValueSelected(suggestion);
                      break;
                    }
                    case "ArrowDown": {
                      if (!event.currentTarget.nextElementSibling) {
                        event.preventDefault();
                      } else {
                        (event.currentTarget.nextElementSibling as HTMLElement).focus();
                      }
                      break;
                    }
                    case "ArrowUp": {
                      if (!event.currentTarget.previousElementSibling) {
                        event.preventDefault();
                      } else {
                        (event.currentTarget.previousElementSibling as HTMLElement).focus();
                      }
                      break;
                    }
                    default: {
                      break;
                    }
                  }
                }}
                onClick={() => {
                  dispatch({ type: actions.reset });
                  dispatch({ type: actions.setSelectedValue, payload: props.suggestions });
                  dispatch({ type: actions.setFilter, payload: suggestion });
                  props.onValueSelected(suggestion);
                }}
              >
                {suggestion}
              </FilteredSuggestionItem>)
            }
          </FilteredSuggestions>
        }
      </Main>
    </ErrorBoundary>
  )
};