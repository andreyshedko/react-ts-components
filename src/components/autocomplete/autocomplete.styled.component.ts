import styled from "styled-components";

export const Main = styled.section`
  font-family: "Open Sans", sans-serif;
  position: relative;
`;

export const FilteredSuggestions = styled.ul`
  position: absolute;
  z-index: 1;
  background-color: #ffffff;
  border: 1px solid #bfbfbf;
  padding-left: 0px;
  top: 30px;
  width: 100%;
`;

export const FilteredSuggestionItem = styled.li`
  padding: 0.5rem;
  cursor: pointer;
  list-style: none;
  box-shadow: 0 0 1px rgba(0,0,0,0.6);
  max-width: 100%;
`;