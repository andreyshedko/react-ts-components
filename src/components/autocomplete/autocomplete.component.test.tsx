import React from "react";

import renderer from 'react-test-renderer';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { AutocompleteComponent } from "./autocomplete.component";

configure({ adapter: new Adapter() });

describe("Autocompplete Component", () => {
    test("should display component", async () => {
        const component = renderer.create(
            <AutocompleteComponent
                suggestions={['Kiev', 'Kirov', 'Moscow', 'Minsk', 'Prague']}
                onValueSelected={jest.fn()}
            />
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

});