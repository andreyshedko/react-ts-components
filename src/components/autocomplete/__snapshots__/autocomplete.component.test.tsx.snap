// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocompplete Component should display component 1`] = `
<section
  className="sc-AxgMl bGQGqg"
>
  <section
    className="sc-AxjAm kMCKjL"
  >
    <input
      className="sc-AxirZ kquce"
      disabled={false}
      maxLength={100}
      minLength={1}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      placeholder=""
      required={true}
      tabIndex={0}
      type="text"
      value=""
    />
  </section>
</section>
`;
